DSpace/DSpace
# Re-enable automated CSRF Protection in Spring Security for Server Webapp

## References
* This issue was discovered by LGTM, as (currently) our `WebSecurityConfiguration` includes `csrf().disabled()`.  See LGTM Rule: https://lgtm.com/rules/1511220334423/
* **MUST BE MERGED WITH ANGULAR PR:** https://github.com/DSpace/dspace-angular/pull/833  (If this PR is merged prior to the Angular PR, then authentication/authorization will no longer work in the Angular UI as every auth request will return a `403 Forbidden`)


## Description
As noted in that [LGTM rule](https://lgtm.com/rules/1511220334423/), Spring Security enables this CSRF protection by default and recommends it "for any request that could be processed by a browser client by normal users".

**Simply put, this PR reenables CSRF protection in Spring Security, and configures it based on [Angular best practices](https://angular.io/guide/http#security-xsrf-protection).**

Previously, we had chosen to *disable* this default CSRF protection because (as originally designed) we planned to make _all authentication checks_ via the `Authorization: Bearer...` Header.  If Authentication checks only occur via Headers, then CSRF protection is unnecessary (as [CSRF only applies to Cookies](https://security.stackexchange.com/questions/62080/is-csrf-possible-if-i-dont-even-use-cookies)).

However, we changed this behavior recently when implementing Shibboleth Authentication in PR #2651.  Shibboleth implementation required us to also support Cookie-based authentication checks.

On current `main`, while we _primarily_ still do authentication checks via that `Authorization` Header, we **now do support authentication checks via Cookie**.  See this newly added IT `testStatusPasswordAuthenticatedWithCookie()` which verifies that: https://github.com/DSpace/DSpace/commit/2d489ad70a1795551b20222ab8b4833607b14ac4

Because of this newly added behavior (and the fact that we plan to store additional session/user info in Cookies, see #2807), we really should have CSRF protection enabled.

**NOTE: After this is accepted, we will need to make minor updates to the REST Contract.**  I promise to make these changes once this has been accepted. 
1. All modification requests (e.g. POST, PUT, PATCH, DELETE) will require CSRF verification (X-XSRF-TOKEN header) to succeed. Otherwise a 403 Forbidden error will be returned.  Read only requests (GET, OPTIONS) do NOT require CSRF verification.
2. Our CSRF verification process aligns with the recommendations of Spring Security + Angular
    * The REST API sends a `XSRF-TOKEN` Cookie in each request.
    * When making a modification request, the Client **must** include a `X-XSRF-TOKEN` Header in the request, with the value of the `XSRF-TOKEN` Cookie in that Header.

## Instructions for Reviewers
Reviewers should:
1. Verify code & tests look good
2. Verify HAL Browser still works (it needed minor updates to support CSRF protection), especially authentication & access restricted endpoints.
3. (Optionally) test it with Shibboleth authentication as well. (However, I don't believe it should be affected.)

List of changes in this PR:
* Re-enabled CSRF protection in `WebSecurityConfiguration`
    * Took the opportunity to restructure `WebSecurityConfiguration` & add more inline comments, as it had gotten confusing.
* Allowed the `X-XSRF-TOKEN` header in `Application` settings
* Added CSRF protection support to all ITs.  See changes to `AbstractControllerIntegrationTest`. Required adding `spring-security-test` as a dependency
* Updated HAL Browser to support XSRF-TOKEN Cookie & X-XSRF-TOKEN Header
    * While making these modifications, I stumbled on a minor XSS security issue in HAL Browser described in https://github.com/mikekelly/hal-browser/pull/97.  So, I applied that fix.
* Minor updates to `AbstractBuilderCleanupUtil` to add in missing Builders.  I hit some strange random test errors that were solved by adding missing Builders into the cleanup list.
* **UPDATE:** This PR has been updated to ensure both the XSRF-TOKEN Cookie **and** the DSpace auth Cookie both have `SameSite=None; Secure`.  This ensures that both cookies can be used/sent cross domain (e.g. if UI is on a different host from backend, or if Shibboleth is on a different host from backend).


## Checklist
